File: python/examples/stream_grpc.py
#!/usr/bin/env python3
"""
gRPC Streaming Example - High-Performance Real-Time Data

Stream trades, orders, L2 book, L4 book, and blocks via gRPC.
gRPC provides lower latency than WebSocket for high-frequency trading.

The SDK:
- Connects to port 10000 automatically
- Passes token via x-token header
- Handles reconnection with exponential backoff
- Manages keepalive pings

Requirements:
    pip install hyperliquid-sdk[grpc]

Usage:
    export ENDPOINT="https://your-endpoint.example.com/TOKEN"
    python stream_grpc.py
"""

import os
import signal
import sys
import time
from datetime import datetime

from hyperliquid_sdk import HyperliquidSDK

ENDPOINT = os.environ.get("ENDPOINT") or os.environ.get("QUICKNODE_ENDPOINT")

if not ENDPOINT:
    print("gRPC Streaming Example")
    print("=" * 60)
    print()
    print("Usage:")
    print("  export ENDPOINT='https://your-endpoint.example.com/TOKEN'")
    print("  python stream_grpc.py")
    sys.exit(1)


def timestamp():
    return datetime.now().isoformat()[11:23]


def main():
    print("=" * 60)
    print("gRPC Streaming Examples")
    print("=" * 60)

    # Example 1: Stream Trades
    print("\nExample 1: Streaming Trades")
    print("-" * 60)

    trade_count = [0]

    # Create SDK client
    sdk = HyperliquidSDK(ENDPOINT)
    grpc = sdk.grpc()

    def on_connect():
        print("[CONNECTED]")

    def on_error(err):
        print(f"[ERROR] {err}")

    def on_trade(data):
        trade_count[0] += 1
        coin = data.get("coin", "?")
        px = float(data.get("px", "0"))
        sz = data.get("sz", "?")
        side = "BUY " if data.get("side") == "B" else "SELL"
        print(f"[{timestamp()}] {side} {sz} {coin} @ ${px:,.2f}")

    grpc.on_connect = on_connect
    grpc.on_error = on_error

    grpc.trades(["BTC", "ETH"], on_trade)
    print("Subscribing to BTC and ETH trades...")

    # Handle Ctrl+C gracefully
    def signal_handler(sig, frame):
        print("\nShutting down gracefully...")
        grpc.stop()
        sys.exit(0)

    signal.signal(signal.SIGINT, signal_handler)

    grpc.start()

    # Wait for trades or timeout
    start = time.time()
    while trade_count[0] < 10 and time.time() - start < 30:
        time.sleep(0.1)

    print(f"\nReceived {trade_count[0]} trades.")
    grpc.stop()

    # Example 2: Stream L2 Book
    print("\nExample 2: Streaming L2 Order Book")
    print("-" * 60)

    l2_count = [0]

    sdk2 = HyperliquidSDK(ENDPOINT)
    grpc2 = sdk2.grpc()

    def on_l2_book(data):
        l2_count[0] += 1
        bids = data.get("bids", [])
        asks = data.get("asks", [])
        if bids and asks:
            best_bid = float(bids[0][0]) if isinstance(bids[0], list) else float(bids[0].get("price", 0))
            best_ask = float(asks[0][0]) if isinstance(asks[0], list) else float(asks[0].get("price", 0))
            spread = best_ask - best_bid
            print(f"[{timestamp()}] ETH: bid=${best_bid:,.2f} ask=${best_ask:,.2f} spread=${spread:.2f}")

    grpc2.l2_book("ETH", on_l2_book, n_levels=10)
    grpc2.start()

    start = time.time()
    while l2_count[0] < 5 and time.time() - start < 15:
        time.sleep(0.1)

    print(f"\nReceived {l2_count[0]} L2 updates.")
    grpc2.stop()

    # Example 3: Stream Blocks
    print("\nExample 3: Streaming Blocks")
    print("-" * 60)

    block_count = [0]

    sdk3 = HyperliquidSDK(ENDPOINT)
    grpc3 = sdk3.grpc()

    def on_block(data):
        block_count[0] += 1
        bn = data.get("block_number", "?")
        print(f"[{timestamp()}] Block #{bn}")

    grpc3.blocks(on_block)
    grpc3.start()

    start = time.time()
    while block_count[0] < 5 and time.time() - start < 20:
        time.sleep(0.1)

    print(f"\nReceived {block_count[0]} blocks.")
    grpc3.stop()

    print()
    print("=" * 60)
    print("Done!")


if __name__ == "__main__":
    main()

File: typescript/examples/stream_grpc.ts
#!/usr/bin/env npx ts-node
// @ts-nocheck
/**
 * gRPC Streaming Example - High-Performance Real-Time Data
 *
 * Stream trades, orders, L2 book, L4 book, and blocks via gRPC.
 * gRPC provides lower latency than WebSocket for high-frequency trading.
 *
 * Usage:
 *     export ENDPOINT="https://your-endpoint.example.com/TOKEN"
 *     npx ts-node stream_grpc.ts
 *
 * The SDK:
 * - Connects to port 10000 automatically
 * - Passes token via x-token header
 * - Handles reconnection with exponential backoff
 * - Manages keepalive pings
 */

import { HyperliquidSDK } from '@quicknode/hyperliquid-sdk';

const ENDPOINT = process.env.ENDPOINT;

if (!ENDPOINT) {
  console.log("gRPC Streaming Example");
  console.log("=".repeat(60));
  console.log();
  console.log("Usage:");
  console.log("  export ENDPOINT='https://your-endpoint.example.com/TOKEN'");
  console.log("  npx ts-node stream_grpc.ts");
  process.exit(1);
}

function timestamp(): string {
  const now = new Date();
  return now.toISOString().slice(11, 23);
}

async function main() {
  console.log("=".repeat(60));
  console.log("gRPC Streaming Examples");
  console.log("=".repeat(60));

  // Example 1: Stream Trades
  console.log("\nExample 1: Streaming Trades");
  console.log("-".repeat(60));

  let tradeCount = 0;

  // Create SDK client
  const sdk = new HyperliquidSDK(ENDPOINT);

  sdk.grpc.onConnect = () => console.log("[CONNECTED]");
  sdk.grpc.onError = (err) => console.log(`[ERROR] ${err.message}`);

  sdk.grpc.trades(["BTC", "ETH"], (data: any) => {
    tradeCount++;
    const coin = data.coin || "?";
    const px = parseFloat(data.px || "0");
    const sz = data.sz || "?";
    const side = data.side === "B" ? "BUY " : "SELL";
    console.log(`[${timestamp()}] ${side} ${sz} ${coin} @ $${px.toLocaleString()}`);

    if (tradeCount >= 5) {
      console.log(`\nReceived ${tradeCount} trades.`);
    }
  });

  console.log("Subscribing to BTC and ETH trades...");

  await sdk.grpc.start();

  // Wait for trades or timeout
  const start = Date.now();
  while (tradeCount < 5 && Date.now() - start < 15000) {
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  sdk.grpc.stop();

  // Example 2: Stream L2 Book
  console.log("\nExample 2: Streaming L2 Order Book");
  console.log("-".repeat(60));

  let l2Count = 0;

  // Create new SDK client for L2
  const sdk2 = new HyperliquidSDK(ENDPOINT);

  sdk2.grpc.l2Book("ETH", (data: any) => {
    l2Count++;
    const bids = data.bids || [];
    const asks = data.asks || [];
    if (bids.length > 0 && asks.length > 0) {
      const bestBid = parseFloat(bids[0][0] || "0");
      const bestAsk = parseFloat(asks[0][0] || "0");
      const spread = bestAsk - bestBid;
      console.log(`[${timestamp()}] ETH: bid=$${bestBid} ask=$${bestAsk} spread=$${spread.toFixed(2)}`);
    }

    if (l2Count >= 3) {
      console.log(`\nReceived ${l2Count} L2 updates.`);
    }
  }, { nLevels: 10 });

  await sdk2.grpc.start();

  const l2Start = Date.now();
  while (l2Count < 3 && Date.now() - l2Start < 10000) {
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  sdk2.grpc.stop();

  // Example 3: Stream Blocks
  console.log("\nExample 3: Streaming Blocks");
  console.log("-".repeat(60));

  let blockCount = 0;

  // Create new SDK client for blocks
  const sdk3 = new HyperliquidSDK(ENDPOINT);

  sdk3.grpc.blocks((data: any) => {
    blockCount++;
    const bn = data.block_number || "?";
    console.log(`[${timestamp()}] Block #${bn}`);

    if (blockCount >= 3) {
      console.log(`\nReceived ${blockCount} blocks.`);
    }
  });

  await sdk3.grpc.start();

  const blockStart = Date.now();
  while (blockCount < 3 && Date.now() - blockStart < 15000) {
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  sdk3.grpc.stop();

  console.log("\n" + "=".repeat(60));
  console.log("Done!");
}

main().catch(console.error);
